scalanet/src/io/iohk/scalanet/peergroup/InMemoryPeerGroup.scala

```diff
@@ -0,0 +1,142 @@
+package io.iohk.scalanet.peergroup
+
+import monix.eval.Task
+import monix.reactive.{MulticastStrategy, Observable}
+import monix.reactive.subjects.ConcurrentSubject
+import monix.execution.Scheduler.Implicits.global
+import InMemoryPeerGroup._
+
+import scala.collection.concurrent.TrieMap
+import scala.util.Random
+
+// For simplicity we can model the network with three main interactions.
+// 1. Register to the network.
+// 2. Disconnect from the network.
+// 3. Send messages.
+class Network[A, M] {
+  private val peers: TrieMap[A, InMemoryPeerGroup[A, M]] = TrieMap()
+
+  def clear(): Unit = peers.clear()
+
+  def register(peer: InMemoryPeerGroup[A, M]): Result[Unit] =
+    if (peers.get(peer.processAddress).isEmpty) {
```

Comment:
While it is unlikely to happen, is there a simple way to do a compare and set? if there is, it is worth using it.

Reply:
The thing is that I am not just interested in comparing and setting. I want to actually raise an error if the comparison fails.